index.d.ts

```diff
@@ -0,0 +1,9 @@
+declare module 'svelte-routing' {
+	import { SvelteComponent } from 'svelte'
+	export const Router: SvelteComponent
+	export const Route: SvelteComponent
+	export const Link: SvelteComponent
```

Comment:
Should we not extend the SvelteComponent type with the allowed props on the Link element?

Reply:
Indeed, i was taking a look at the repos sent by @mefechoel alongside with Svelte types, and discovered a [`SvelteComponentTyped` interface](https://github.com/sveltejs/svelte/blob/391455c58e8cd6de8008c661e0f7f1efa7671e43/src/runtime/internal/dev.ts#L150), correct me if i'm wrong, but maybe it's the most suitable solution at the moment due to the fact that allows the inclusion of props, events and even slots through generics, what you think?

Any feedback would be greatly appreciated

Reply:
Nice! I didn't know about `SvelteComponentTyped`. A very similar custom type is what is used in the examples I linked. I agree, this should be the way to go, as it comes from the official svelte repo.

Reply:
As a note when typing slots: the name of the unnamed slot, as in `Route` is `default`. So the slot type would be something like `{ default: { location: Location; params: Params } }`. Took me a while to figure that out :'D

Reply:
How exactly would `Location` and `Params` be typed? i couldn't find anything insightful about this on the docs besides a brief explanation of the latter one:

> Potential path parameters are passed back to the parent using props, so they can be exposed to the slot template using `let:params`

And a paragraph about `getProps` mentioning `location`:

> A function that returns an object that will be spread on the underlying anchor element's attributes. The first argument given to the function is an object with the properties `location`, `href`, `isPartiallyCurrent`, `isCurrent`. Look at the [`NavLink` component in the example project setup](https://github.com/EmilTholin/svelte-routing/blob/master/example/src/components/NavLink.svelte) to see how you can build your own link components with this.

I presume that they could be typed like:

```typescript
interface RouteSlots {
  default: {
    location: string
    params: Record<string, any>
  }
}
```

Reply:
Location is made up from `window.location` plus a `state` property. However when the memory history implementation is used it just has the `pathname`, `search` and `state` properties. I'm not sure why the `hash` property is missing. This leaves us with a common interface like this:

```ts
interface RouterLocation {
  pathname: string;
  search: string;
  hash?: string;
  state: { [k: string | number]: unknown };
}
```

Since the params are parsed from strings the interface would look like this:

```ts
interface Params {
  [param: string]: string;
}
```

Reply:
Just a little correction to your first example:

```typescript
state: {
    [k in string | number]: unknown
} 
```

> The parameter type of a index signature cannot be a union type, a mapped object type must be used instead `ts(1337)`

Reply:
Oops, you're right of course, thanks :)